### PR TITLE
nix: convert to a flake based structure

### PR DESCRIPTION
**Pull request by eadwu**
In progress.

Basically a upstreaming of usable parts from `ngi-nix/mangaki` while preserving as much of the original structure.

Not really sure what `poetryShell` is for?. The VM is also missing it's attribute `vm` and file `./mangaki-vm.nix`. Same with `wheeledShell` and `sourceShell`, I can't tell where they come from.

### CODE REVIEW

File: nix/modules/mangaki.nix [hunk 1]
```diff
@@ -258,19 +299,44 @@ in
...
+
+        StateDirectory = "mangaki";
+        StateDirectoryMode = "0750";
+        WorkingDirectory = "/var/lib/mangaki";
+      };
+
+      # TODO: django-admin runserver bugs out looking like it fails to parse bash
+      script = ''
+        django-admin migrate
+        python ${pkgs.mangaki.src}/mangaki/manage.py runserver
```
Comment (eadwu) [inline nix/modules/mangaki.nix:324]: Not really familiar with django so I'm not sure `django-admin runserver` is failing.

**Comment by RaitoBezarius** [comment]
First, thanks a lot for opening a PR here!! I really appreciate it.

> Not really sure what poetryShell is for?. The VM is also missing it's attribute vm and file ./mangaki-vm.nix. Same with wheeledShell and sourceShell, I can't tell where they come from.

That was mostly testing stuff, it's not really relevant.

I'll review it ASAP. I'm exciting to have it!

File: nix/modules/mangaki.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nix/modules/mangaki.nix:333]: I'll take a look to it

**Comment by Elarnon** [comment]
I don't know much about using Nix, and so I can't comment on the rest of the PR, but it looks like this is downloading code directly from `https://github.com/edolstra/flake-compat/archive/master.tar.gz` without any sort of verification on what is downloaded.

This doesn't seem in line with the reproducibility goals of the project, in addition of technically being a security risk.

**Comment by RaitoBezarius** [comment]
> I don't know much about using Nix, and so I can't comment on the rest of the PR, but it looks like this is downloading code directly from `https://github.com/edolstra/flake-compat/archive/master.tar.gz` without any sort of verification on what is downloaded.
> 
> This doesn't seem in line with the reproducibility goals of the project, in addition of technically being a security risk.

This is true, in practice, it'll do a "trust on first-use" (because the store will cache the tarball with some sort of TTL).
We just have to pin the sha256 to enforce reproducibility.
And I'm getting `1vi8bb024mfl266nax7zpnppnzp0976kacqxnvfdgsmny0sj1ka3` according to `nix-prefetch-url`

**Comment by Elarnon** [comment]
> This is true, in practice, it'll do a "trust on first-use" (because the store will cache the tarball with some sort of TTL).
> We just have to pin the sha256 to enforce reproducibility.
> And I'm getting `1vi8bb024mfl266nax7zpnppnzp0976kacqxnvfdgsmny0sj1ka3` according to `nix-prefetch-url`

While this is true re: security risks, given that the URL is a tarball for the `master` branch of a git repository, this will probably not solve reproducibility issues.

**Comment by RaitoBezarius** [comment]
> > This is true, in practice, it'll do a "trust on first-use" (because the store will cache the tarball with some sort of TTL).
> > We just have to pin the sha256 to enforce reproducibility.
> > And I'm getting `1vi8bb024mfl266nax7zpnppnzp0976kacqxnvfdgsmny0sj1ka3` according to `nix-prefetch-url`
> 
> While this is true re: security risks, given that the URL is a tarball for the `master` branch of a git repository, this will probably not solve reproducibility issues.

Right. It is possible to lock the tarball "properly": https://nixos.wiki/wiki/Flakes ("Using flakes project from legacy Nix").
FWIW, this is only a library shim to support the Flakes mechanism, so I was not really that concerned, but this is a correct point.

**Comment by eadwu** [comment]
Pinned the repository using the flake-based method, currently at `cecfd08`.

File: nix/modules/mangaki.nix (hunk 1 above)
Comment (RaitoBezarius) [reply nix/modules/mangaki.nix:333]: Okay, I know why: `django-admin runserver` will proceed to a restart of its own executable, but the current exec is a bash file and not a Python file, so it does not parse properly and throws a syntax error.

The fix would be to patch Django's reloader, I guess.

Comment (RaitoBezarius) [reply nix/modules/mangaki.nix:333]: We can keep the things as-is and put a comment on the top of it.

**Comment by RaitoBezarius** [comment]
Thanks @eadwu for this PR, I'll close this one and keep track of the rest in #707

**Pull request closed without merging**